src/abstractdataframe/nest.jl

```diff
@@ -0,0 +1,390 @@
+   1 │ b       3×2 DataFrame 
+   2 │ a       2×2 DataFrame 
+   3 │ c       1×2 DataFrame 
+
+julia> n2.zx
+3-element Vector{DataFrame}:
+ 3×2 DataFrame
+ Row │ z     x     
+     │ Char  Int64 
+─────┼─────────────
+   1 │ a         1
+   2 │ e         5
+   3 │ f         6
+ 2×2 DataFrame
+ Row │ z     x     
+     │ Char  Int64 
+─────┼─────────────
+   1 │ b         2
+   2 │ c         3
+ 1×2 DataFrame
+ Row │ z     x     
+     │ Char  Int64 
+─────┼─────────────
+   1 │ d         4
+
+julia> n3 = nest(groupby(df, :id), :x => "x", [:y, :z] => "yz")
+3×3 DataFrame
+ Row │ id      x              yz
+     │ String  DataFrame      DataFrame     
+─────┼──────────────────────────────────────
+   1 │ b       3×1 DataFrame  3×2 DataFrame 
+   2 │ a       2×1 DataFrame  2×2 DataFrame 
+   3 │ c       1×1 DataFrame  1×2 DataFrame 
+
+julia> n3.x
+3-element Vector{DataFrame}:
+ 3×1 DataFrame
+ Row │ x     
+     │ Int64 
+─────┼───────
+   1 │     1
+   2 │     5
+   3 │     6
+ 2×1 DataFrame
+ Row │ x     
+     │ Int64 
+─────┼───────
+   1 │     2
+   2 │     3
+ 1×1 DataFrame
+ Row │ x     
+     │ Int64 
+─────┼───────
+   1 │     4
+
+julia> n3.yz
+3-element Vector{DataFrame}:
+ 3×2 DataFrame
+ Row │ y      z    
+     │ Int64  Char 
+─────┼─────────────
+   1 │    11  a
+   2 │    15  e
+   3 │    16  f
+ 2×2 DataFrame
+ Row │ y      z    
+     │ Int64  Char 
+─────┼─────────────
+   1 │    12  b
+   2 │    13  c
+ 1×2 DataFrame
+ Row │ y      z    
+     │ Int64  Char
+─────┼─────────────
+   1 │    14  d
+"""
+nest(gdf::GroupedDataFrame, cols::Pair{<:Any, <:AbstractString}...) =
+    combine(gdf, (sdf -> (; Symbol(dst) => select(sdf, index(sdf)[src]))
+                  for (src, dst) in cols)...)
+nest(gdf::GroupedDataFrame, cols::Pair{<:Any, Symbol}...) =
+    combine(gdf, (sdf -> (; dst => select(sdf, index(sdf)[src]))
+                  for (src, dst) in cols)...)
+nest(gdf::GroupedDataFrame) = nest(gdf, valuecols(gdf) => :data)
+
+"""
+    unnest(df::AbstractDataFrame, src::ColumnIndex...;
+           cols::Union{Symbol, AbstractVector{Symbol},
+                       AbstractVector{<:AbstractString}}=:setequal,
+           promote::Bool=true,
+           makeunique::Bool=false, flatten::Bool=true)
+
+Unnest one or more columns `src` into multiple columns. The newly created
+columns are stored at the end of the data frame (and the `src` column is
+dropped).
+
+Each `src` column must contain a `NamedTuple`, a `DataFrameRow`, a
+`Tables.AbstractRow`, or Tables.jl table.
+
+`cols` (default `:setequal`) and `promote` (default `true`) keyword arguments
+have the same meaning as in [`push!`](@ref).
+
+If `makeunique=false` (the default) produced column names must be unique.
+If `makeunique=true` then duplicate column names will be suffixed with `_i`
+(`i` starting at `1` for the first duplicate).
+
+If `flatten=true` (the default) then newly created columns are flattened
+using [`flatten`](@ref) with `scalar=Missing` keyword argument.
+
```

Comment:
Missing examples

Reply:
yes - I first want to discuss the API.

Reply:
If we were to discuss the case that I think might be most problematic is:
```
julia> df = DataFrame(x=[(a="aa", b="bb"), (a="pp", b="qq")])
2×1 DataFrame
 Row │ x
     │ NamedTup…
─────┼──────────────────────
   1 │ (a = "aa", b = "bb")
   2 │ (a = "pp", b = "qq")

julia> unnest(df, :x, flatten=false)
2×2 DataFrame
 Row │ a       b      
     │ String  String 
─────┼────────────────
   1 │ aa      bb
   2 │ pp      qq

julia> unnest(df, :x)
4×2 DataFrame
 Row │ a     b    
     │ Char  Char 
─────┼────────────
   1 │ a     b
   2 │ a     b
   3 │ p     q
   4 │ p     q
```
so maybe we should allow for passing `scalar` kwarg here also or not allow to flatten at all (but expect it to be a separate step).

Reply:
What would be the output of that operation with `scalar`?

Reply:
I'm not clear on why `unnest` supports unpacking single-row cells. I expected it to work on Table-valued cells, not row-valued ones. That seems like a different operation.

Reply:
> What would be the output of that operation with scalar?

If we allowed passing `scalar` user could write:

```
julia> unnest(df, :x, scalar=String)
2×2 DataFrame
 Row │ a       b      
     │ String  String 
─────┼────────────────
   1 │ aa      bb
   2 │ pp      qq
```
and it would not get flattened.

---

> I expected it to work on Table-valued cells, not row-valued ones. That seems like a different operation.

I assume you have experience with dplyr, where there is no distinction between table-values and row-values. In JuliaData ecosystem there is such a distinction, and both operations are useful, and are essentially `unnest` (and we need to make a design that fits JuliaData ecosystem, not just mimics dplyr). Here is an example of the difference:

```
julia> df = DataFrame(data=[(a=1, b=2), (a=3, c=4)])
2×1 DataFrame
 Row │ data
     │ NamedTup…      
─────┼────────────────
   1 │ (a = 1, b = 2)
   2 │ (a = 3, c = 4)

julia> unnest(df, :data, flatten=false, cols=:union)
2×3 DataFrame
 Row │ a      b        c       
     │ Int64  Int64?   Int64?  
─────┼─────────────────────────
   1 │     1        2  missing 
   2 │     3  missing        4

julia> df = DataFrame(data=[(a=1:2, b=3:4), (a=5:6, c=7:8)])
2×1 DataFrame
 Row │ data
     │ NamedTup…
─────┼────────────────────
   1 │ (a = 1:2, b = 3:4)
   2 │ (a = 5:6, c = 7:8)

julia> unnest(df, :data, flatten=true, cols=:union)
4×3 DataFrame
 Row │ a      b        c       
     │ Int64  Int64?   Int64?  
─────┼─────────────────────────
   1 │     1        3  missing 
   2 │     2        4  missing 
   3 │     5  missing        7
   4 │     6  missing        8

julia> unnest(df, :data, flatten=false, cols=:union)
2×3 DataFrame
 Row │ a          b           c
     │ UnitRang…  UnitRang…?  UnitRang…?
─────┼───────────────────────────────────
   1 │ 1:2        3:4         missing
   2 │ 5:6        missing     7:8
```
As you can see for row-like input only one operation makes sense (`flatten=false`), for table-like input two operations make sense (`flatten=true`, and `flatten=false`). All three options have been asked for by the users, so I propose design that supports all of them. We could indeed define two separate functions (one for row-data and one for table-data), but it would be almost identical (with the difference that for row data there would be no `flatten` kwarg, but for table-data still `flatten` would be needed).

My initial thought if we need to provide `flatten` keyword argument at all as it is the same as calling `flatten` in the last step. But maybe we should for convenience support it, and then maybe it is useful to allow passing `scalar` keyword argument.

Reply:
I don't like `flatten=`. In general, rather than `f_and_maybe_g(;g=true)` I prefer `g(f())`. Especially here where so much meaning depends on that kwarg.

---

In some other issue, a function `union` was preferred over the symbol `:union`. Perhaps here, too?

Reply:
> I don't like `flatten=`.

I also have this feeling. Maybe we could define `unnest` that works as `flatten=true` and e.g. `spread` that works as `flatten=false` now.

In practice `unnest` then would call `spread` and then `flatten` (the convenience of this is that it would automatically know which columns should be flattened).

---
Regarding `cols` kwarg it supports the following operations:
```
    •  If cols == :setequal (this is the default) then row must contain exactly the same columns as df (but possibly in a different order).

    •  If cols == :orderequal then row must contain the same columns in the same order (for AbstractDict this option requires that keys(row) matches
       propertynames(df) to allow for support of ordered dicts; however, if row is a Dict an error is thrown as it is an unordered collection).

    •  If cols == :intersect then row may contain more columns than df, but all column names that are present in df must be present in row and only
       they are used to populate a new row in df.

    •  If cols == :subset then the behavior is like for :intersect but if some column is missing in row then a missing value is pushed to df.

    •  If cols == :union then columns missing in df that are present in row are added to df (using missing for existing rows) and a missing value is
       pushed to columns missing in row that are present in df.
```
I would keep them as is, as changing this would be breaking (+ some of them are not easily expressed as functions).

Reply:
`flatten=false` is in my opinion, in the context of DataFrames.jl more natural. I know that dplyr users expect `flatten=true`, but a natural use-case for data-frames is:
1. you get a source JSON file
2. your read it in into a data frame and get something like:
```
julia> df = DataFrame(col=[(a="x", b=1), (a="y", c=[1,2])])
2×1 DataFrame
 Row │ col
     │ NamedTup…
─────┼───────────────────────
   1 │ (a = "x", b = 1)
   2 │ (a = "y", c = [1, 2])
```
3. you want that tuned into:
```
julia> df = DataFrame(a=["x", "y"], b=[1, missing], c=[missing, [1,2]])
2×3 DataFrame
 Row │ a       b        c
     │ String  Int64?   Array…?
─────┼──────────────────────────
   1 │ x             1  missing
   2 │ y       missing  [1, 2]
```
I would say that this is the most commonly Julia-native pattern needed. If named-tuples were homogenous it would just be `select(df, :col => AsTable)`. Therefore, originally users requested adding an option in `AsTable` allowing to specify the `cols` and `promote` arguments from `push!`.

However, I am hesitant with complicating operation specification syntax even more, therefore for this special case I preferred a separate function.

`flatten=true` is rarely needed in practice because `GroupedDataFrame` in practice gives the user almost an equivalent interface.

Reply:
I see. Though I would find it surprising if similar to the example you gave in your previous post `unnest(DataFrame(data=[(a=1:2, b=3:4), (a=5:6, b=7:8)]), :data)` gave a single-row result. At least when a valid Tables.jl table is passed (in particular a `DataFrame`) it's natural that each of its rows gives a row in the result.

Maybe we can treat objects in a different way depending on whether they are tables or rows by default, while providing an argument for users to request explicitly the behavior they want? That would be useful in particular if you happen to have a named tuples of vectors which should not be treated as a table.

Reply:
I have thought about this, but then there is a corner case of users mixing tables and non-tables (in one column or across columns).

If we feel `unnest` is useful (I think it is) then - as I have proposed - maybe indeed we need two separate functions: one doing flattening, and other not doing it?

Reply:
OK - I have thought about it more and now I have a clear mind about it on an abstract level.

So essentially if you have a nested structure in a cell of a data frame you can ask for the following operations:
1. spread it into multiple rows (this is current `flatten`)
2. spread it into multiple columns if it defines valid column names (we do not have it now; this is the `flatten=false` option in proposed `unnest`)
3. perform both operations 1. and 2. (which makes sense only for tables; this is the `flatten=true` option in proposed `unnest`)

(note that technically some data structures support both 1 and 2: e.g. `NamedTuple` could be spread into multiple rows or multiple columns conceptually)

Given this - maybe indeed we need to add two functions. The verbose names would be `unnest_as_row` and `unnest_as_table` (they are verbose, so probably we should not use them, but they show the intention).

Reply:
I decided to split `unnest` and `expand`